src/modules-arch/module-descriptor.js

```diff
@@ -0,0 +1,221 @@
+   * function like _ => import('./path).
+   *
+   * @param {ModuleDescriptor.constructor[]} [opts.dependencies]
+   * Constructors of modules the module depends on.
+   *
+   * @param {ModuleDescriptor.constructor[]} [opts.allowedSubmodules]
+   * Constructors of modules the module could use as submodules.
+   *
+   * @param {ModuleDescriptor[]} [opts.submodules]
+   * Modules that the module will use as submodules.
+   * Cannot contain any entries except those described in
+   * opts.allowedSubmodules.
+   *
+   * @param {ModuleDescriptor.constructor[]} [opts.incompatibles]
+   * Constructors of modules that cannot be used simultaneously with the module.
+   *
+   *
+   * @param {Boolean} [opts.isCorporateOnly]
+   * If `true` the component will be accessible only by corporate accounts
+   *
+   */
+  constructor (opts = {}) {
+    // will be assigned automatically while the component renders
+    this._createdComponentUid = -1
+    this._createdComponent = {}
+
+    const {
+      dependencies = [],
+      allowedSubmodules = [],
+      submodules = [],
+      incompatibles = [],
+      importComponent = null,
+      importStore = null,
+    } = opts
+
+    if (typeof importComponent !== 'function') {
+      throw new Error(`${this.constructor.name}: no importComponent provided`)
+    }
+
+    this._importComponent = this._rememberUidAfterImport(importComponent)
+
+    if (importStore) {
+      this._importStore = importStore
+    }
+
+    this._dependencies = dependencies
+    this._incompatibles = incompatibles
+    this._allowedSubmodules = allowedSubmodules
+    this._isCorporateOnly = opts.isCorporateOnly || false
+
+    this.validateSubmodules(submodules)
+    this._submodules = submodules
+  }
+
+  get importComponent () { return this._importComponent }
+  get importStore () { return this._importStore }
+  get dependencies () { return this._dependencies }
+  get allowedSubmodules () { return this._allowedSubmodules }
+  get submodules () { return this._submodules }
+  get incompatibles () { return this._incompatibles }
+  get isAccessible () {
+    return this._isCorporateOnly
+      ? store.getters[vuexTypes.isAccountCorporate]
+      : true
+  }
+  get createdComponentUid () { return this._createdComponentUid }
+  get createdComponent () { return this._createdComponent }
+  get isComponentCreated () { return this._createdComponentUid >= 0 }
+
+  _rememberUidAfterImport (importFn) {
+    return async _ => {
+      if (this.isComponentCreated) {
+        return importFn()
+      }
+
+      const imported = await importFn()
+      const componentDefinition = Object.values(imported)
+        .find(item => item.render)
+
+      if (!componentDefinition) {
+        throw new Error(`${this.constructor.name}: cannot find imported component definition`)
+      }
+
+      if (!componentDefinition.beforeCreate) {
+        componentDefinition.beforeCreate = []
+      }
+
+      const self = this
```

Comment:
it's 2k19 and we have arrow functions now

Reply:
Added a comment why we cannot use arrow function here